In [1]:
import json
import pandas as pd

# JSON Data Import Stack Overflow Solution

https://stackoverflow.com/questions/21058935/python-json-loads-shows-valueerror-extra-data

# Receipts Data

In [2]:
receipts = []
for line in open('receipts.json', 'r'):
    receipts.append(json.loads(line))

In [3]:
receipts = pd.DataFrame(receipts)

In [4]:
receipts.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}        500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}        150.0  {'$date': 1609601083000}   
2                       NaN            5  {'$date': 1609632000000}   
3  {'$date': 1609687534000}          5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}          5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus totalSpent                    userId  
0             FINISHED      26.00  5ff1e1eacfcf6c399c274ae6  
1             FINISHED      11.00  5ff1e194b6a9d73a3a9f1052  
2             REJECTED      10.00  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED      28.00  5ff1e1eacfcf6c399c274ae6  
4             FINISHED       1.00  5ff1e194b6a9d73a3a9f1052

In [5]:
receipts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      1119 non-null   object 
 1   bonusPointsEarned        544 non-null    float64
 2   bonusPointsEarnedReason  544 non-null    object 
 3   createDate               1119 non-null   object 
 4   dateScanned              1119 non-null   object 
 5   finishedDate             568 non-null    object 
 6   modifyDate               1119 non-null   object 
 7   pointsAwardedDate        537 non-null    object 
 8   pointsEarned             609 non-null    object 
 9   purchaseDate             671 non-null    object 
 10  purchasedItemCount       635 non-null    float64
 11  rewardsReceiptItemList   679 non-null    object 
 12  rewardsReceiptStatus     1119 non-null   object 
 13  totalSpent               684 non-null    object 
 14  userId                  

## Takeaways
1. There are a total of 1119 receipts in the data provided
2. There are less values in the pointsAwardedDate (537) column than pointsEarned(609). Missing data?

## Why are there less values in the pointsAwardedDate (537) column than pointsEarned(609)?

1. pointAwardedDate - the date points were awarded for the transaction. 
2. pointsEarned - the number of points earned for the receipt. 
3. Conculsion: There might be missing data. They should be equal. Need to explore further. 

In [97]:
receipts.columns

Index(['_id', 'bonusPointsEarned', 'bonusPointsEarnedReason', 'createDate',
       'dateScanned', 'finishedDate', 'modifyDate', 'pointsAwardedDate',
       'pointsEarned', 'purchaseDate', 'purchasedItemCount',
       'rewardsReceiptItemList', 'rewardsReceiptStatus', 'totalSpent',
       'userId'],
      dtype='object')

In [98]:
# created a new dataframe for the receipts that do not have a point Awarded Date

noDate_receipts = receipts[receipts['pointsAwardedDate'].isna()]
noDate_receipts.head()

_id  bonusPointsEarned  \
2   {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
12  {'$oid': '5ff1e1b60a7214ada100055c'}              150.0   
15  {'$oid': '5ff1e1e90a7214ada1000569'}                NaN   
28  {'$oid': '5ff1e1d40a7214ada1000562'}                NaN   
62  {'$oid': '5ff4a4ca0a7214ada10005d0'}              750.0   

                              bonusPointsEarnedReason  \
2                          All-receipts receipt bonus   
12  Receipt number 5 completed, bonus point schedu...   
15                                                NaN   
28                                                NaN   
62  Receipt number 1 completed, bonus point schedu...   

                  createDate               dateScanned finishedDate  \
2   {'$date': 1609687537000}  {'$date': 1609687537000}          NaN   
12  {'$date': 1609687478000}  {'$date': 1609687478000}          NaN   
15  {'$date': 1609687529000}  {'$date': 1609687529000}          NaN   
28  {'$date': 1609687508000}  {'$date': 1609687508000}          NaN   
62  {'$date': 1609868490000}  {'$date': 1609868490000}          NaN   

                  modifyDate pointsAwardedDate pointsEarned  \
2   {'$date': 1609687542000}               NaN            5   
12  {'$date': 1609687478000}               NaN       8850.0   
15  {'$date': 1609687529000}               NaN          NaN   
28  {'$date': 1609687508000}               NaN          NaN   
62  {'$date': 1609868492000}               NaN        750.0   

                purchaseDate  purchasedItemCount  \
2   {'$date': 1609632000000}                 1.0   
12  {'$date': 1612365878000}                10.0   
15                       NaN                 0.0   
28  {'$date': 1509321600000}                 3.0   
62  {'$date': 1599955200000}                 2.0   

                               rewardsReceiptItemList rewardsReceiptStatus  \
2   [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
12  [{'barcode': '034100573065', 'description': 'M...              FLAGGED   
15  [{'needsFetchReview': True, 'needsFetchReviewR...              FLAGGED   
28  [{'deleted': True, 'description': 'DELETED ITE...             REJECTED   
62  [{'description': 'mueller austria hypergrind p...             REJECTED   

   totalSpent                    userId  
2       10.00  5ff1e1f1cfcf6c399c274b0b  
12     290.00  5ff1e194b6a9d73a3a9f1052  
15       0.00  5ff1e1e9b6a9d73a3a9f10f6  
28       3.00  5ff1e194b6a9d73a3a9f1052  
62      34.96  5fbc35711d967d1222cbfefc

In [99]:
# reward Status Finished

noDate_receipts[noDate_receipts['rewardsReceiptStatus'] == 'FINISHED' ]

_id  bonusPointsEarned  \
123  {'$oid': '5ff774200a7214ada1000638'}                NaN   
174  {'$oid': '5ff8c8880a7214adca000008'}                NaN   
186  {'$oid': '5ffc8cc20a720f05c5000026'}                NaN   
206  {'$oid': '5ffc8cb40a720f05c5000025'}                NaN   

    bonusPointsEarnedReason                createDate  \
123                     NaN  {'$date': 1610052640000}   
174                     NaN  {'$date': 1610139784000}   
186                     NaN  {'$date': 1610350626000}   
206                     NaN  {'$date': 1610350612000}   

                  dateScanned              finishedDate  \
123  {'$date': 1610052640000}  {'$date': 1610052640000}   
174  {'$date': 1610139784000}  {'$date': 1610139785000}   
186  {'$date': 1610350626000}  {'$date': 1610368681000}   
206  {'$date': 1610350612000}  {'$date': 1610368616000}   

                   modifyDate pointsAwardedDate pointsEarned  \
123  {'$date': 1610052640000}               NaN          0.0   
174  {'$date': 1610139785000}               NaN          0.0   
186  {'$date': 1610368681000}               NaN          0.0   
206  {'$date': 1610368616000}               NaN          0.0   

                 purchaseDate  purchasedItemCount  \
123  {'$date': 1609977600000}                 2.0   
174  {'$date': 1610064000000}                 2.0   
186  {'$date': 1610287200000}                 1.0   
206  {'$date': 1610287200000}                 1.0   

                                rewardsReceiptItemList rewardsReceiptStatus  \
123  [{'brandCode': 'WINGSTOP', 'description': '12 ...             FINISHED   
174  [{'brandCode': 'WINGSTOP', 'description': '12 ...             FINISHED   
186  [{'brandCode': 'WINGSTOP', 'description': '12 ...             FINISHED   
206  [{'brandCode': 'WINGSTOP', 'description': '12 ...             FINISHED   

    totalSpent                    userId  
123       6.18  5ff7741e04929111f6e90902  
174       6.18  5ff8c88604929111f6e913b7  
186       3.09  5ffc8cc104929111f6e922a3  
206       3.09  5ffc8cb304929111f6e9229d

## Takeaways

1. Data Quality Issue: Missing Data
        
        In the pointsAwardedDate column, there are some receipts that received points, but the date the points were awareded are not recorded. As seen in receipts 2,12, and 62 above, they were all awarded points (5, 8850, and 750) but the date that the points were received are omitted.

2. The receipts in this dataframe with the receipt status of "FINISHED" all have points Earned = 0

### Additional Receipt Data Takeaways

    1. There are a total of 1119 receipts in data provided

    2. pointsAwardedDate column: 
    why is it less than pointsEarned? There might be missing data. They should be equal
    potential data quality issue: missing data
    
    3. pointsEarned column: 
    Does every receipt receive points as advertised on app? If so, there is missing data or some customers did not receive points. 
    potential data quality issue: missing data

    4. purchaseDate column: 
    Why do some receipts not have purchase dates? On the app, the date is required on a receipt?
    potential data quality issue: missing data
    
    5. purchasedItemCount column: 
    It is OK to have Null values in this column. i.e. Restaurants will have 0 item count but earn points. 
    
    6. rewardsReceiptItemList column: 
    Only 679/1119 values
    Barcode is missing for some receipts. Need to explore further.
    
    7. rewardsReceiptStatus column: 
    Need to explore further

## Questions to Answer
  
    1. What is done with receipts with 0 items? Do they have barcodes? i.e. A receipt for a new restaurant
    2. What are the different receipt statuses?

### What is done with receipts with 0 items? Do they have barcodes? i.e. A receipt for a new restaurant

In [48]:
# No Item receipts data
noItem_receipts = receipts[receipts['purchasedItemCount'] == 0]
noItem_receipts

_id  bonusPointsEarned  \
15   {'$oid': '5ff1e1e90a7214ada1000569'}                NaN   
81   {'$oid': '5ff4ce3c0a720f05230005c4'}                NaN   
141  {'$oid': '5ff73be90a720f052300060a'}                NaN   
144  {'$oid': '5ff794600a7214ada1000647'}                NaN   
175  {'$oid': '5ff8da570a720f05c5000015'}                NaN   
179  {'$oid': '5ff8da7d0a720f05c500001c'}                NaN   
282  {'$oid': '6000d4aa0a720f05f3000072'}                NaN   
396  {'$oid': '6009eb000a7214ada2000003'}              250.0   
424  {'$oid': '600aff160a720f053500000c'}              500.0   
544  {'$oid': '60132b890a7214ad50000013'}                NaN   
559  {'$oid': '60145a540a720f05f8000116'}                NaN   
642  {'$oid': '601830cd0a720f05f800034f'}                NaN   
656  {'$oid': '60182f390a7214ad500002b8'}                NaN   
681  {'$oid': '60189c920a7214ad2800003a'}                NaN   
727  {'$oid': '601b32160a7214ad28000221'}                NaN   

                               bonusPointsEarnedReason  \
15                                                 NaN   
81                                                 NaN   
141                                                NaN   
144                                                NaN   
175                                                NaN   
179                                                NaN   
282                                                NaN   
396  Receipt number 3 completed, bonus point schedu...   
424  Receipt number 2 completed, bonus point schedu...   
544                                                NaN   
559                                                NaN   
642                                                NaN   
656                                                NaN   
681                                                NaN   
727                                                NaN   

                   createDate               dateScanned  \
15   {'$date': 1609687529000}  {'$date': 1609687529000}   
81   {'$date': 1609879100000}  {'$date': 1609879100000}   
141  {'$date': 1610038249000}  {'$date': 1610038249000}   
144  {'$date': 1609996096000}  {'$date': 1609996096000}   
175  {'$date': 1610144343000}  {'$date': 1610144343000}   
179  {'$date': 1610144381000}  {'$date': 1610144381000}   
282  {'$date': 1610667178000}  {'$date': 1610667178000}   
396  {'$date': 1611262720000}  {'$date': 1611262720000}   
424  {'$date': 1611333398000}  {'$date': 1611333398000}   
544  {'$date': 1611869065000}  {'$date': 1611869065000}   
559  {'$date': 1611946580000}  {'$date': 1611946580000}   
642  {'$date': 1612198093000}  {'$date': 1612198093000}   
656  {'$date': 1612197689000}  {'$date': 1612197689000}   
681  {'$date': 1612225682000}  {'$date': 1612225682000}   
727  {'$date': 1612395030000}  {'$date': 1612395030000}   

                 finishedDate                modifyDate  \
15                        NaN  {'$date': 1609687529000}   
81                        NaN  {'$date': 1609879100000}   
141                       NaN  {'$date': 1610038249000}   
144                       NaN  {'$date': 1610039297000}   
175                       NaN  {'$date': 1610144344000}   
179                       NaN  {'$date': 1610144381000}   
282                       NaN  {'$date': 1610667178000}   
396  {'$date': 1611262746000}  {'$date': 1611262755000}   
424  {'$date': 1611333421000}  {'$date': 1611333433000}   
544                       NaN  {'$date': 1611869065000}   
559                       NaN  {'$date': 1611946580000}   
642                       NaN  {'$date': 1612198093000}   
656                       NaN  {'$date': 1612197690000}   
681                       NaN  {'$date': 1612225683000}   
727                       NaN  {'$date': 1612395031000}   

            pointsAwardedDate pointsEarned              purchaseDate  \
15                        NaN          NaN                       NaN   
81                        NaN          NaN        

### Takeaways
1. For most of the 0 item receipts, 0 points earned and many have rwardsReceiptStatus of flagged or rejected. 
2. There are 2 receipts (index 396 and 424) that generated points, but even these receipts do not have barcoes.
3. Need further clarification on how 0 item receipts are accounted for.

### What are the different receipt status?

In [66]:
status_receipts = receipts.groupby(receipts['rewardsReceiptStatus'])

In [67]:
status_receipts.groups.keys()

dict_keys(['FINISHED', 'FLAGGED', 'PENDING', 'REJECTED', 'SUBMITTED'])

# Brands Data

In [100]:
brands = []
for line in open('brands.json', 'r'):
    brands.append(json.loads(line))

In [50]:
brands = pd.DataFrame(brands)

In [51]:
brands.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name topBrand                      brandCode  
0  test brand @1612366101024    False                            NaN  
1                  Starbucks    False                      STARBUCKS  
2  test brand @1612366146176    False  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051    False  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827    False  TEST BRANDCODE @1612366146827

In [52]:
brands.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           1167 non-null   object
 1   barcode       1167 non-null   object
 2   category      1012 non-null   object
 3   categoryCode  517 non-null    object
 4   cpg           1167 non-null   object
 5   name          1167 non-null   object
 6   topBrand      555 non-null    object
 7   brandCode     933 non-null    object
dtypes: object(8)
memory usage: 73.1+ KB


### Takeaways

    1. There are a total of 1167 brands in data
    
    2. Each item has a name, barcode, and cpg (reference to consumer packaged good). 
        Examples of consumer packaged goods include food, beverages, cosmetics, and cleaning products.
    
    3. category column: 
        category (1012) out of 1167 entries
        Do some brands not fall into a category? 
        potential data quality issue = missing data
        
     4. Are there any restaurants in the data? If so, what is the category, categoryCode, and cpg?

### Different Categories for Brands

In [62]:
# Different Cateogries
category_brands = brands.groupby(brands['category'])

In [65]:
category_brands.groups.keys()

dict_keys(['Baby', 'Baking', 'Beauty', 'Beauty & Personal Care', 'Beer Wine Spirits', 'Beverages', 'Bread & Bakery', 'Breakfast & Cereal', 'Candy & Sweets', 'Canned Goods & Soups', 'Cleaning & Home Improvement', 'Condiments & Sauces', 'Dairy', 'Dairy & Refrigerated', 'Deli', 'Frozen', 'Grocery', 'Health & Wellness', 'Household', 'Magazines', 'Outdoor', 'Personal Care', 'Snacks'])

### Takeaway
1. Restaurant category is missing from the brands data

# User Data

In [75]:
users = []
for line in open('users.json', 'r'):
    users.append(json.loads(line))

In [76]:
users = pd.DataFrame(users)

In [77]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


In [78]:
users.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

### Takeaways

    1. There are 495 users in the data

    2. _id column:
    Looks like there are duplicate users. Need to explore further. 

## Check for duplicate values

In [80]:
users.columns

Index(['_id', 'active', 'createdDate', 'lastLogin', 'role', 'signUpSource',
       'state'],
      dtype='object')

In [94]:
# Return the rows with duplicate user IDs

dupl_users = users.loc[users['_id'].duplicated(),:]
dupl_users

_id  active               createdDate  \
1    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
4    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
5    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
8    {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
..                                    ...     ...                       ...   
490  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
491  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
492  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
493  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   
494  {'$oid': '54943462e4b07e684157a532'}    True  {'$date': 1418998882381}   

                    lastLogin         role signUpSource state  
1    {'$date': 1609687537858}     consumer        Email    WI  
2    {'$date': 1609687537858}     consumer        Email    WI  
4    {'$date': 1609687537858}     consumer        Email    WI  
5    {'$date': 1609687537858}     consumer        Email    WI  
8    {'$date': 1609687537858}     consumer        Email    WI  
..                        ...          ...          ...   ...  
490  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
491  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
492  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
493  {'$date': 1614963143204}  fetch-staff          NaN   NaN  
494  {'$date': 1614963143204}  fetch-staff          NaN   NaN  

[283 rows x 7 columns]

## Takeaways
1. There are duplicate user IDs in the users data. 
2. Looks like the duplicate user IDs were created on similar dates.
3. Some of the duplicate user IDs were created by fetch-staff. Maybe test accounts?

In [95]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


## Final Takeaways

1. Additional signUpSource column: 
       1. This column is missing from the data schema
       2. This column could be missing data. There are only 447 rows of data out of 495. Are these the user who select prefer not to specify?
